jwst/flatfield/flat_field.py

```diff
@@ -187,17 +189,40 @@ def apply_flat_field(science, flat):
     # For CubeModel science data, apply flat to each integration
     if isinstance(science, datamodels.CubeModel):
         for integ in range(science.data.shape[0]):
-            # Flatten data and error arrays
+            # Flatten data
             science.data[integ] /= flat_data
-            science.err[integ] /= flat_data
+
+            # Update the variances using BASELINE algorithm
+            science.var_poisson[integ] /= flat_data**2
+            science.var_rnoise[integ] /= flat_data**2
+            science.var_flat[integ] = science.data[integ]**2 / flat_data**2 * flat_err**2
+
+            # Propogate flat and flat variance into the science err array
+            science.err[integ] = np.sqrt(
+                science.var_poisson[integ] +
+                science.var_rnoise[integ] +
+                science.var_flat[integ]
+                )
+
             # Combine the science and flat DQ arrays
             science.dq[integ] = np.bitwise_or(science.dq[integ], flat_dq)
 
     # For 2D ImageModel science data, apply flat to entire arrays
     else:
         # Flatten data and error arrays
         science.data /= flat_data
-        science.err /= flat_data
+
+        # Update the variances using BASELINE algorithm
+        science.var_poisson /= flat_data**2
+        science.var_rnoise /= flat_data**2
+        science.var_flat = science.data**2 / flat_data**2 * flat_err**2
```

Comment:
Here and above could you compute `flat_data * *2` once, assign it to a variable and reuse it.